content/cpi-api-v2.md

```diff
@@ -0,0 +1,115 @@
+
+- [BOSH components](bosh-components.md) and its example component interaction diagram
+- [CLI v2 architecture doc](https://github.com/cloudfoundry/bosh-cli/blob/master/docs/architecture.md#deploy-command-flow) and [`bosh create-env` flow](https://github.com/cloudfoundry/bosh-init/blob/master/docs/init-cli-flow.png) where calls to the CPI are marked as `cloud`.
+
+Examples of API request and response:
+
+- [Building a CPI: RPC - Request](https://bosh.io/docs/build-cpi.html#request)
+- [Building a CPI: RPC - Response](https://bosh.io/docs/build-cpi.html#response)
+
+
+Library:
+
+- Ruby: `bosh-cpi-ruby` gem [v2.5.0](https://github.com/cloudfoundry/bosh-cpi-ruby/releases/tag/v2.5.0)
+- GoLang: `bosh-cpi-go` [library](https://github.com/cppforlife/bosh-cpi-go)
+
+---
+
+## Glossary {: #glossary }
+
+- **cloud ID** is an ID (string) that the Director uses to reference any created infrastructure resource; typically CPI methods return cloud IDs and later receive them. For example AWS CPI's `create_vm` method would return `i-f789df` and `attach_disk` would take it.
+
+- **cloud_properties** is a hash that can be specified for several objects (resource pool, disk pool, stemcell, network) to provide infrastructure specific settings to the CPI for that object. Only CPIs know the meaning of its contents. For example resource pool's `cloud_properties` for AWS can specify `instance_type`:
+
+```yaml
+resource_pools:
+- name: large_machines
+  cloud_properties: {instance_type: r3.8xlarge}
+```
+
+## Methods
+
+- All V1 contracts must still be supported. See [CPI API V1](cpi-api-v1.md).
+- To differentiate V2 calls the caller needs to pass in `"api_version": 2` in the context
+
+### Reference Table (Based on each component version)
+* Reg./reg. : Registry
+
+| Director | CPI | Stemcell  | Should update Reg.   | Add agent setting to Iaas `user-metadata`   |
+|----------|-----|-----------|----------------------|---|
+| 1  | 1  | 1  | Update Reg.  | Dont add agent setting to iaas  |
+| 1  | 1  | 2  | Update Reg.  | Dont add agent setting to iaas  |
+| 1  | 2  | 2  | Update Reg.  | Dont add agent setting to iaas  |
+| 2  | 2  | 2  | DON'T add anything to reg.   | Yes, agent will read `user-metadata` and not call reg.  |
+| 1  | 2  | 1  | Update Reg.  | Dont add agent setting to iaas  |
+| 2  | 2  | 1  | Update Reg.  | Dont add agent setting to iaas (no agent support)  |
+| 2  | 1  | 1  | Update Reg. (cpi will by default update reg.)  | Dont add agent setting to iaas |
+| 2  | 1  | 2  | Update Reg. (cpi will by default update reg.)  | Dont add agent setting to iaas |
+
+### Implementation
+
+API version 2 of CPIs will differ from version 1 by the following:
+
+ - Director will send cpi api_version based on info response from for all CPI calls
+ - Director will also send stemcell `api_version` for all CPI calls
+ - Director will be expecting V2 responses IF
+   - CPI info call provide max supported version as >=2 AND
+   - Stemell api_version is >=2 AND
```

Comment:
Is stemcell version >=2 really a requirement for Director to expect a CPI API v2 response? I think the response for CPI API v2 should be clearly defined and always be the same. The director can decide to ignore parts of it, because the stemcell doesn't support it, but that's a different story. Having different method signatures might not be easily feasible or even desired in all languages, right?

Reply:
We are working on this PR right now in order to improve the content and structure.